Feature: configurable redirect

Ardnived opened the pull request:
This contains the changes in PR #67, but allows further customisation, and also applied the customised redirect to all failures, instead of just token reuse. This also addresses #19 and #21 

 - Sign out redirection can be configured with `sign_out_redirect_path`
 - Failure redirection with `failure_redirect_path`
 - Success redirection with `success_redirect_path`
 - Success redirection can also be set with the `url` query parameter.

Success redirection has this priority
`params[:url]` > `reset_passwordless_redirect_location` > `Passwordless.success_redirect_path`

In theory I've also set it up so that you can override the functions that fetch these redirects for more complicated/customisable redirection, but I haven't tested that.

All tests have been updated and pass.

mikker commented:
This is just wonderful! Thank you for doing this 🙏

1. I'm not sure about `params[:url]` – can we make up a more descriptive name? Like, `destination_url` or `return_url` or something similar.
2. Can you add all this to the README docs? Don't have to be perfect, a first draft is fine 😅

I would be happy to merge this 💙💚💛💜❤️

Ardnived commented:
@mikker No problem, I'll change `params[:url]` to `params[:destination_url]` and add something to the README this evening, and then update the PR.

Ardnived commented on app/controllers/passwordless/sessions_controller.rb:
```diff
@@ -42,28 +42,40 @@ def create
+      return Passwordless.success_redirect_path unless Passwordless.redirect_back_after_sign_in
+
+      redirect_url = reset_passwordless_redirect_location!(authenticatable_class)
+      params[:url] || redirect_url || Passwordless.success_redirect_path
```
A fair point, I've added some code which strips out the host and schema from any URI that is passed in, so that it will just be used as a path redirect.

Ardnived commented:
@mikker I've changed the redirect query parameter to `destination_path`, and taken @JoeSouthan 's suggestion to make it more secure.

I updated the README, and took the liberty of making a few other changes to the README to hopefully make a bit easier to navigate.

JoeSouthan commented on app/controllers/passwordless/sessions_controller.rb:
```diff
@@ -42,28 +42,51 @@ def create
+        query_redirect_path << uri.path
+        query_redirect_path << '?' << uri.query if uri.query.present?
+        query_redirect_path << '#' << uri.fragment if uri.fragment.present?
+        query_redirect_path = query_redirect_path.string
```
I think this could possibly be simplified to:

```ruby
  def query_redirect_path
    query_redirect_uri = URI(params[:destination_path])
    query_redirect_uri.to_s if query_redirect_uri.host == URI(request.url).host
  rescue URI::InvalidURIError
    nil
  end
```

Inspired by https://github.com/rails/rails/blob/66cabeda2c46c582d19738e1318be8d59584cc5b/actionpack/lib/action_controller/metal/redirecting.rb#L127-L131

What do you think?

Ardnived replied:
Absolutely, I've pushed a change for this and updated tests.

Ardnived commented:
Simplified the query param redirect based on @JoeSouthan 's suggestion.

mikker commented on test/controllers/passwordless/sessions_controller_test.rb:
```diff
@@ -154,7 +192,7 @@ class << User
       follow_redirect!
 
-      assert_equal "/", path
+      assert_equal Passwordless.success_redirect_path, path
```
No need to change these if they already succeed.
```suggestion
      assert_equal "/", path
```

mikker commented on test/controllers/passwordless/sessions_controller_test.rb:
```diff
@@ -176,7 +214,7 @@ class << User
 
       assert_equal 200, status
-      assert_equal "/", path
+      assert_equal Passwordless.sign_out_redirect_path, path
```
```suggestion
      assert_equal "/", path
```

mikker commented on test/controllers/passwordless/sessions_controller_test.rb:
```diff
@@ -191,7 +229,7 @@ class << User
       assert_nil session[Helpers.session_key(user.class)]
       assert_equal 200, status
-      assert_equal "/", path
+      assert_equal Passwordless.failure_redirect_path, path
```
```suggestion
      assert_equal "/", path
```

mikker commented on test/controllers/passwordless/sessions_controller_test.rb:
```diff
@@ -214,7 +252,7 @@ class << User
       follow_redirect!
       assert_equal 200, status
-      assert_equal "/", path
+      assert_equal Passwordless.failure_redirect_path, path
```
```suggestion
      assert_equal "/", path
```

mikker commented on test/controllers/passwordless/sessions_controller_test.rb:
```diff
@@ -229,7 +267,7 @@ class << User
 
       assert_equal 200, status
-      assert_equal "/", path
+      assert_equal Passwordless.sign_out_redirect_path, path
```
```suggestion
      assert_equal "/", path
```

mikker commented on test/integration/navigation_test.rb:
```diff
@@ -16,7 +16,7 @@ class NavigationTest < ActionDispatch::IntegrationTest
     follow_redirect!
     assert_equal 200, status
-    assert_equal "/", path
+    assert_equal Passwordless.failure_redirect_path, path
```
```suggestion
    assert_equal "/", path
```

mikker commented on test/controllers/passwordless/sessions_controller_test.rb:
```diff
@@ -116,7 +116,7 @@ class << User
 
       assert_equal 200, status
-      assert_equal "/", path
+      assert_equal Passwordless.success_redirect_path, path
```
```suggestion
      assert_equal "/", path
```

mikker commented on test/controllers/passwordless/sessions_controller_test.rb:
```diff
@@ -104,7 +104,7 @@ class << User
 
       assert_equal 200, status
-      assert_equal "/", path
+      assert_equal Passwordless.success_redirect_path, path
```
```suggestion
      assert_equal "/", path
```

stuyam commented:
Glad to see this pr was approved, this will really help me with #70. Let me know if there is anything I can do to help move it forward. Looking forward to it, thanks @Ardnived & @mikker! ❤️

stuyam commented on app/controllers/passwordless/sessions_controller.rb:
```diff
@@ -42,28 +42,47 @@ def create
+
+    def passwordless_query_redirect_path
+      query_redirect_uri = URI(params[:destination_path])
+      query_redirect_uri.to_s if query_redirect_uri.host.nil? || query_redirect_uri.host == URI(request.url).host
```
Very very very smart!

Ardnived commented:
I merged the suggestions. Feel free to merge the PR as far as I'm concerned.

mikker commented:
💙💚💛💜❤️

mikker commented:
Just released in `0.9.0`

stuyam commented:
Legendary work! This pr really takes this package to the next level! 🥇
